src/model/mod.rs

```diff
@@ -392,15 +397,328 @@ pub fn calculate_ratings(
+
+            let gwr = &gwrs[g_idx];
+
+            let mut p = 1;
+            for s in s_clone {
+                let scores = p_scores.entry(s.player_id).or_insert(Vec::new());
+                let misses = p_misses.entry(s.player_id).or_insert(Vec::new());
+                let accs = p_accs.entry(s.player_id).or_insert(Vec::new());
+                let placement = p_placement.entry(s.player_id).or_insert(Vec::new());
+                let gplayed = p_gplayed.entry(s.player_id).or_insert(0);
+                let gwon = p_gwon.entry(s.player_id).or_insert(0);
+                let glost = p_glost.entry(s.player_id).or_insert(0);
+
+                p_ids.push(s.player_id);
+                scores.push(s.score);
+                misses.push(s.misses);
+                accs.push(s.accuracy_standard);
+                placement.push(p);
+                *gplayed += 1;
+
+                let won = player_won_game(&s.player_id, &gwr);
+                if won {
+                    *gwon += 1;
+                } else {
+                    // Ties are technically losses in this case, we can figure this out later.
+                    *glost += 1;
+                }
+
+                p += 1;
+            }
+
+            g_idx += 1;
+        }
+
+        p_ids = p_ids.into_iter().unique().collect();
+
+        let winning_roster = mwr.winner_roster.clone();
+        let losing_roster = mwr.loser_roster.clone();
+
+        for p_id in p_ids {
+            let won = winning_roster.contains(&p_id);
+            res.push(PlayerMatchStats {
+                player_id: p_id,
+                match_id: m.id,
+                won,
+                average_score: mean_i64(p_scores.entry(p_id).or_insert(Vec::new())),
+                average_misses: mean_i32(p_misses.entry(p_id).or_insert(Vec::new())),
+                average_accuracy: mean_f64(p_accs.entry(p_id).or_insert(Vec::new())),
+                average_placement: mean_i32(p_placement.entry(p_id).or_insert(Vec::new())),
+                games_won: *p_gwon.entry(p_id).or_insert(0),
+                games_lost: *p_glost.entry(p_id).or_insert(0),
+                games_played: *p_gplayed.entry(p_id).or_insert(0),
+                teammate_ids: if won {
+                    winning_roster
+                        .clone()
+                        .iter()
+                        .filter(|x| **x != p_id)
+                        .map(|x| *x)
+                        .collect()
+                } else {
+                    losing_roster
+                        .clone()
+                        .iter()
+                        .filter(|x| **x != p_id)
+                        .map(|x| *x)
+                        .collect()
+                },
+                opponent_ids: if won {
+                    losing_roster.clone()
+                } else {
+                    winning_roster.clone()
+                }
+            });
+        }
+    }
+
+    res
+}
+
+fn mean_i32(numbers: &[i32]) -> f64 {
```

Comment:
Quality: all these can be somewhat easily turned into a single `fn mean<T>(numbers: &[T]) -> f64`, but that requires changing the type of score into `i32` due to the way type conversion constrains work

Should be fine temporarily as it isn't a breaking issue but it should be addressed just out of code quality concerns

Reply:
Implemented and converted all scores to i32